File: src/Scanner.Core/Checks/EmptyLinkCheck.cs
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Checks for links that lack accessible names.
/// </summary>
public sealed class EmptyLinkCheck : ICheck
{
    private static readonly Regex LinkRegex = new("<a(?<attrs>[^>]*)>(?<content>.*?)</a>", RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);

    /// <inheritdoc />
    public string Id => "empty-link";

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        var issues = new List<Issue>();
        var elementIds = AccessibleNameUtilities.CollectElementIds(context.Content);

        foreach (Match match in LinkRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            if (!IsLink(attrs))
            {
                continue;
            }

            var content = match.Groups["content"].Value;
            if (AccessibleNameUtilities.HasTextContent(content))
            {
                continue;
            }

            var ariaLabel = AttributeParser.GetAttributeValue(attrs, "aria-label");
            if (AccessibleNameUtilities.HasAriaLabel(ariaLabel))
            {
                continue;
            }

            var ariaLabelledBy = AttributeParser.GetAttributeValue(attrs, "aria-labelledby");
            if (AccessibleNameUtilities.HasValidAriaLabelledBy(ariaLabelledBy, elementIds))
            {
                continue;
            }

            var title = AttributeParser.GetAttributeValue(attrs, "title");
            if (AccessibleNameUtilities.HasTitle(title))
            {
                continue;
            }

            var line = TextUtilities.GetLineNumber(context.Content, match.Index);
            issues.Add(new Issue(rule.Id, Id, context.FilePath, line, "Link has no accessible name.", match.Value));
        }

        return issues;
    }

    private static bool IsLink(string attributes)
    {
        if (TextUtilities.ContainsAttribute(attributes, "href"))
        {
            return true;
        }

        var role = AttributeParser.GetAttributeValue(attributes, "role");
        return string.Equals(role, "link", StringComparison.OrdinalIgnoreCase);
    }
}

File: node_engine/src/checks/EmptyLinkCheck.js
const { getAttributeValue } = require("./AttributeParser");
const { getLineNumber, getLineNumberForSnippet, containsAttribute } = require("./TextUtilities");
const {
  collectElementIds,
  collectElementIdsFromDocument,
  hasAriaLabel,
  hasValidAriaLabelledBy,
  hasTitle,
  hasTextContent
} = require("./AccessibleNameUtilities");
const { hasImageAltText, hasImageAltTextFromElement } = require("./MissingLinkTextCheck");

const linkRegex = /<a(?<attrs>[^>]*)>(?<content>[\s\S]*?)<\/a>/gi;

const isLink = (attributes) => {
  if (containsAttribute(attributes, "href")) {
    return true;
  }

  const role = getAttributeValue(attributes, "role");
  return Boolean(role && role.toLowerCase() === "link");
};

const EmptyLinkCheck = {
  id: "empty-link",
  applicableKinds: ["html", "htm", "cshtml", "razor"],
  run(context, rule) {
    const issues = [];
    if (context.document?.querySelectorAll) {
      const elementIds = collectElementIdsFromDocument(context.document);
      const candidates = Array.from(context.document.querySelectorAll("a, [role=\"link\"]"));
      const sourceAnchors = context.content ? Array.from(context.content.matchAll(linkRegex)) : [];
      let sourceIndex = 0;

      for (const link of candidates) {
        const isAnchor = link.tagName?.toLowerCase() === "a";
        const sourceMatch = isAnchor ? sourceAnchors[sourceIndex++] : null;
        const role = link.getAttribute("role");
        const hasHref = link.hasAttribute("href");
        if (!(hasHref || (role && role.toLowerCase() === "link"))) {
          continue;
        }

        if (hasTextContent(link.textContent || "")) {
          continue;
        }

        if (hasImageAltTextFromElement(link)) {
          continue;
        }

        const ariaLabel = link.getAttribute("aria-label");
        if (hasAriaLabel(ariaLabel)) {
          continue;
        }

        const labelledBy = link.getAttribute("aria-labelledby");
        if (hasValidAriaLabelledBy(labelledBy, elementIds)) {
          continue;
        }

        const title = link.getAttribute("title");
        if (hasTitle(title)) {
          continue;
        }

        const evidence = sourceMatch ? sourceMatch[0] : link.outerHTML;
        const matchIndex = sourceMatch?.index ?? -1;
        issues.push({
          ruleId: rule.id,
          checkId: EmptyLinkCheck.id,
          filePath: context.filePath,
          line: matchIndex >= 0
            ? getLineNumber(context.content, matchIndex)
            : getLineNumberForSnippet(context.content, evidence),
          message: "Link has no accessible name.",
          evidence
        });
      }

      return issues;
    }

    const elementIds = collectElementIds(context.content);

    for (const match of context.content.matchAll(linkRegex)) {
      const attrs = match.groups?.attrs ?? "";
      if (!isLink(attrs)) {
        continue;
      }

      const content = match.groups?.content ?? "";
      if (hasTextContent(content)) {
        continue;
      }

      if (hasImageAltText(content)) {
        continue;
      }

      const ariaLabel = getAttributeValue(attrs, "aria-label");
      if (hasAriaLabel(ariaLabel)) {
        continue;
      }

      const labelledBy = getAttributeValue(attrs, "aria-labelledby");
      if (hasValidAriaLabelledBy(labelledBy, elementIds)) {
        continue;
      }

      const title = getAttributeValue(attrs, "title");
      if (hasTitle(title)) {
        continue;
      }

      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: EmptyLinkCheck.id,
        filePath: context.filePath,
        line,
        message: "Link has no accessible name.",
        evidence: match[0]
      });
    }

    return issues;
  }
};

module.exports = { EmptyLinkCheck, isLink };
